File: practice/ArrayQueue.py
class ArrayQueue:
	def __init__(self, size = 8):
		self._capacity = size
		self._array = [None for _ in range(self._capacity)]
		self._front = self._rear = 0
		self._size = 0

	def _resize(self, new_capacity):
		new_array = [None for i in range(new_capacity)]
		new_front = 0
		new_rear = 0
		while self._rear != self._front:
			new_array[new_front] = self._array[self._rear]
			self._rear = (self._rear + 1) % self._capacity
			new_front = (new_front + 1) % new_capacity
		self._capacity = new_capacity
		self._array = new_array
		self._front = new_front
		self._rear = new_rear
		return f"Array resized to {new_capacity}"


	def enqueue(self, value):
		if (self._front + 1) % self._capacity == self._rear:
			self._resize(self._capacity * 2)

		self._array[self._front] = value
		self._front = (self._front + 1) % self._capacity
		self._size += 1
		

	def dequeue(self):
		if (self._array[self._rear] == None):
			print("queue is empty\n")
		else:
			self._array[self._rear] = None
			self._rear = (self._rear + 1) % self._capacity  
			self._size -= 1


	def display(self):
		temp = self._rear

		while temp != self._front:
			print(self._array[temp], end=" ")
			temp = (temp + 1) % self._capacity
		print()


	def empty(self):
		return self._size == 0


	def full(self):
		return self._capacity - 1 == self._size

if __name__ == "__main__":
	import doctest
	import time
	start = time.time()
	print("Testing...")
	doctest.testmod()
	print("Tested status - 200 OK")
	print(f"Time to test: {round(time.time() - start, 10)} ms")

File: practice/ArrayQueue.java
class ArrayQueue{
	private int size;
	private int capacity;
	private int front = 0;
	private int rear = 0;
	private int queue[];

	ArrayQueue(int c){
		capacity = c;
		queue = new int[c];
	}

	ArrayQueue(){
		capacity = 8;
		queue = new int[8];
	}


	public void enqueue(int val){
		if (queue[front] == 0){
			queue[front] = val;
			front = (front + 1) % capacity;
			size++;
			System.out.println("Element added: " + val);
		}
		else{
			System.out.println("Size is not sufficient!");
		}
	}


	public void dequeue(){
		if (queue[rear] == 0){
			System.out.println("Queue is empty!");
		}
		else{
			queue[rear] = 0;
			rear = (rear + 1) % capacity;
			size--;
		}
	}


	public void display(){

		int temp = rear;
		while (temp != front){
			System.out.print(queue[temp] + " ");
			temp = (temp + 1) % capacity;
		}
		System.out.println();
	}


	public static void main(String[] args){
		// Test the object here.
	}
}